### PR TITLE
Add provision to set the AP LAN Ports to Bridge/NAT

### PR DESCRIPTION
**Pull request by chaitanyagodavarthi**
1. Add provision to tag with vlan when set to Bridge mode (br-wan).
2. Add provision to set the local IP address in NAT (br-lan)
3. Display the Ethernet port status(up/down) and negotiated rate.
4. Add provision to set lan eth port to bridge/nat.

Signed-off-by: Chaitanya Godavarthi <chaitanya.kiran@netexperience.com>

### CODE REVIEW

**Review by netexp-ravi — COMMENTED** [review]
Made some specific comments.

File: feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_state.c [hunk 1]
```diff
@@ -100,6 +103,246 @@ int l3_device_split(char *l3_device, struct iface_info *info)
...
+struct eth_port_state {
+	char ifname[8];
+	char state[8];
+	char speed[8];
+	char duplex[16];
+	char bridge[8];
+};
+
+struct eth_port_state wanport;
+struct eth_port_state lanport[MAX_ETH_PORTS];
```
[the remark this answers is not in the document]
Comment (chaitanyagodavarthi) [reply feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_state.c:121]: done

File: feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_state.c [hunk 2]
```diff
@@ -100,6 +103,246 @@ int l3_device_split(char *l3_device, struct iface_info *info)
...
+		len = read(fd, eth_state->duplex, sizeof(eth_state->duplex) -1);
+
+		if (len < 0)
+			snprintf(eth_state->duplex, sizeof(eth_state->duplex), "none");
+		else
+			eth_state->duplex[len-1] = '\0';
+
+		close(fd);
+	}
+	LOGI("Kiran:%s:eth_state->duplex=%s", eth, eth_state->duplex);
```
[the remark this answers is not in the document]
Comment (chaitanyagodavarthi) [reply feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_state.c:271]: done

File: feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_state.c [hunk 3]
```diff
@@ -100,6 +103,246 @@ int l3_device_split(char *l3_device, struct iface_info *info)
...
+					 lanport[i].state, lanport[i].speed, lanport[i].duplex);
+
+				STRSCPY(state->eth_ports[cnt+i], port_status);
+				cnt++;
+				state->eth_ports_len = cnt;
+			}
+		}
+	}
+
+	LOGI("Kiran:%s:eth_ports_len=%d", state->if_name, state->eth_ports_len);
```
[the remark this answers is not in the document]
Comment (chaitanyagodavarthi) [reply feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_state.c:343]: done

File: feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_config.c [hunk 4]
```diff
@@ -345,6 +366,101 @@ static int wifi_inet_conf_add(struct schema_Wifi_Inet_Config *iconf)
...
+				    iconf->if_name, "interface") != 0)
+			return -1;
+
+		/* Add the resulting modified blob to uci */
+		if (blob_to_uci_section(uci, "network",
+				    iconf->if_name,
+				    "interface", dup.head,
+				    &network_param,
+				    del.head) != 0)
+			return -1;
```
[the remark this answers is not in the document]
Comment (chaitanyagodavarthi) [reply feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_config.c:456]: done

File: patches/0063-ubus-notify-when-ethernet-ports-change-state.patch [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
++ void device_set_ifindex(struct device *dev, int ifindex)
++Index: netifd-2019-08-05-5e02f944/ubus.c
++===================================================================
++--- netifd-2019-08-05-5e02f944.orig/ubus.c
+++++ netifd-2019-08-05-5e02f944/ubus.c
++@@ -817,6 +817,31 @@ netifd_dump_status(struct interface *ifa
++ 		netifd_add_interface_errors(&b, iface);
++ }
++ 
+++void ubus_link_state_notify(struct device *dev, bool state)  //kiran
```
[the remark this answers is not in the document]
Comment (chaitanyagodavarthi) [reply patches/0063-ubus-notify-when-ethernet-ports-change-state.patch]: done

**Review by netexp-ravi — APPROVED** [review]
Seems good. Thanks for taking care of the earlier comments.

File: feeds/wlan-ap/opensync/files/bin/check_wan_link.sh [hunk 6]
```diff
@@ -1,6 +1,21 @@
...
+	json_select ..
+json_select ..
+
+[ -n "$ifname" ] || {
+	ifname=$(uci get network.wan.ifname)
+	ifname=${ifname%% *}
+}
+
+
+if="$ifname"
```
Comment (ArifAlam) [inline feeds/wlan-ap/opensync/files/bin/check_wan_link.sh:18]: Have you tried using network_find_wan() from /lib/functions/network.sh?

File: feeds/wlan-ap/opensync/files/etc/hotplug.d/net/30-bridge [hunk 7]
```diff
@@ -1,14 +1,28 @@
...
+	json_select "wan"
+		json_get_vars ifname
+	json_select ..
+json_select ..
+
+[ -n "$ifname" ] || {
+	ifname=$(uci get network.wan.ifname)
+	ifname=${ifname%% *}
+}
+
```
Comment (ArifAlam) [inline feeds/wlan-ap/opensync/files/etc/hotplug.d/net/30-bridge:18]: Same comment as before: will network_find_wan() /lib/functions/network.sh accomplish the same thing?

File: feeds/wlan-ap/opensync/files/etc/hotplug.d/net/60-dynamic-vlan [hunk 8]
```diff
@@ -1,5 +1,20 @@
...
+	json_select "wan"
+		json_get_vars ifname
+	json_select ..
+json_select ..
+
+[ -n "$ifname" ] || {
+	ifname=$(uci get network.wan.ifname)
+	ifname=${ifname%% *}
+}
+
```
Comment (ArifAlam) [inline feeds/wlan-ap/opensync/files/etc/hotplug.d/net/60-dynamic-vlan:17]: Same comment as before: will network_find_wan() /lib/functions/network.sh accomplish the same thing?

File: feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_config.c [hunk 9]
```diff
@@ -386,6 +502,13 @@ static void callback_Wifi_Inet_Config(ovsdb_update_monitor_t *mon,
 		}
 
 		wifi_inet_conf_add(iconf);
+		system("cp /etc/config/network /etc/config/bkp-network");
```
Comment (ArifAlam) [inline feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_config.c:505]: Save temp stuff in /tmp dir

File: feeds/wlan-ap/opensync/files/bin/check_wan_link.sh (hunk 6 above)
Comment (chaitanyagodavarthi) [reply feeds/wlan-ap/opensync/files/bin/check_wan_link.sh:18]: It uses ubus call network.interface dump internally, I don't see any info containing the ethX interfaces in it.

File: feeds/wlan-ap/opensync/files/etc/hotplug.d/net/30-bridge (hunk 7 above)
Comment (chaitanyagodavarthi) [reply feeds/wlan-ap/opensync/files/etc/hotplug.d/net/30-bridge:18]: It uses ubus call network.interface dump internally, I don't see any info containing the ethX interfaces in it.

File: feeds/wlan-ap/opensync/files/etc/hotplug.d/net/60-dynamic-vlan (hunk 8 above)
Comment (chaitanyagodavarthi) [reply feeds/wlan-ap/opensync/files/etc/hotplug.d/net/60-dynamic-vlan:17]: It uses ubus call network.interface dump internally, I don't see any info containing the ethX interfaces in it.

File: feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_config.c (hunk 9 above)
Comment (chaitanyagodavarthi) [reply feeds/wlan-ap/opensync/src/platform/openwrt/src/netifd/src/wifi_inet_config.c]: done

**Review by ArifAlam — APPROVED** [review]

**Review by ArifAlam — APPROVED** [review]

**Comment by ArifAlam** [comment]
Merged to pending. Thanks.

**Pull request closed without merging**